File: backend/app/routers/transcription.py
import asyncio

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.transcription import (
    MergeRequest,
    SegmentResponse,
    SegmentUpdate,
    SplitRequest,
    TranscriptionRequest,
)
from app.services.segment_service import SegmentService
from app.services.transcription_service import TranscriptionService

router = APIRouter()


@router.post("/start", response_model=list[SegmentResponse])
async def start_transcription(
    data: TranscriptionRequest,
    db: AsyncSession = Depends(get_db),
):
    """Запустити транскрипцію аудіо файлу."""
    from app.services.audio_file_service import AudioFileService

    # Verify audio file exists
    af_service = AudioFileService(db)
    audio_file = await af_service.get_by_id(data.audio_file_id)
    if not audio_file:
        raise HTTPException(status_code=404, detail="Аудіо файл не знайдено")

    # Clear existing segments for this audio file
    seg_service = SegmentService(db)
    existing = await seg_service.get_by_audio_file(data.audio_file_id)
    for seg in existing:
        await seg_service.delete(seg.id)

    # Transcribe
    try:
        ts = TranscriptionService()
        raw_segments = await asyncio.to_thread(
            ts.transcribe,
            audio_file.file_path,
            language=data.language,
            model_size=data.model_size,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Помилка транскрипції: {str(e)}")

    if not raw_segments:
        raise HTTPException(status_code=400, detail="Транскрипція не виявила мовлення")

    # Save segments to database
    segments = await seg_service.create_batch(
        project_id=data.project_id,
        audio_file_id=data.audio_file_id,
        segments=raw_segments,
    )

    # Update project status
    from app.services.project_service import ProjectService
    await ProjectService(db).update_status(data.project_id, "transcribing")

    return segments


@router.get("/segments/{project_id}", response_model=list[SegmentResponse])
async def list_segments(project_id: str, db: AsyncSession = Depends(get_db)):
    """Отримати всі сегменти проєкту."""
    service = SegmentService(db)
    return await service.get_by_project(project_id)


@router.put("/segments/{segment_id}", response_model=SegmentResponse)
async def update_segment(
    segment_id: str,
    data: SegmentUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Оновити текст або включення сегменту."""
    service = SegmentService(db)
    segment = await service.update(segment_id, text=data.text, included=data.included)
    if not segment:
        raise HTTPException(status_code=404, detail="Сегмент не знайдено")
    return segment


@router.delete("/segments/{segment_id}")
async def delete_segment(segment_id: str, db: AsyncSession = Depends(get_db)):
    """Видалити сегмент."""
    service = SegmentService(db)
    deleted = await service.delete(segment_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Сегмент не знайдено")
    return {"success": True}


@router.post("/segments/merge", response_model=SegmentResponse)
async def merge_segments(data: MergeRequest, db: AsyncSession = Depends(get_db)):
    """Об'єднати кілька сегментів."""
    service = SegmentService(db)
    merged = await service.merge(data.segment_ids)
    if not merged:
        raise HTTPException(status_code=400, detail="Не вдалося об'єднати сегменти")
    return merged


@router.post("/segments/{segment_id}/split", response_model=list[SegmentResponse])
async def split_segment(
    segment_id: str,
    data: SplitRequest,
    db: AsyncSession = Depends(get_db),
):
    """Розділити сегмент на два."""
    service = SegmentService(db)
    result = await service.split(segment_id, data.split_time)
    if not result:
        raise HTTPException(status_code=400, detail="Не вдалося розділити сегмент")
    return list(result)


@router.get("/models")
async def available_models():
    """Доступні моделі Whisper та рекомендації для GPU."""
    return {
        "models": [
            {"id": "tiny", "size_mb": 75, "vram_mb": 400, "speed": "fastest", "quality": "low"},
            {"id": "base", "size_mb": 145, "vram_mb": 500, "speed": "fast", "quality": "medium"},
            {"id": "small", "size_mb": 488, "vram_mb": 1000, "speed": "medium", "quality": "good"},
            {"id": "medium", "size_mb": 1500, "vram_mb": 2500, "speed": "slow", "quality": "high"},
            {"id": "large-v3", "size_mb": 3000, "vram_mb": 4500, "speed": "slowest", "quality": "best"},
        ],
        "recommended": "small",
        "note": "RTX 3050 (4GB): small рекомендовано, medium можливо",
    }

File: frontend/src/api/transcription.ts
import { api } from './client'

export interface Segment {
  id: string
  project_id: string
  audio_file_id: string
  start_time: number
  end_time: number
  text: string
  text_edited: boolean
  included: boolean
  created_at: string
}

export interface WhisperModel {
  id: string
  size_mb: number
  vram_mb: number
  speed: string
  quality: string
}

export const transcriptionApi = {
  start: (projectId: string, audioFileId: string, modelSize = 'small', language?: string) =>
    api
      .post<Segment[]>('/transcription/start', {
        project_id: projectId,
        audio_file_id: audioFileId,
        model_size: modelSize,
        language: language || null,
      })
      .then((r) => r.data),

  getSegments: (projectId: string) =>
    api.get<Segment[]>(`/transcription/segments/${projectId}`).then((r) => r.data),

  updateSegment: (segmentId: string, data: { text?: string; included?: boolean }) =>
    api.put<Segment>(`/transcription/segments/${segmentId}`, data).then((r) => r.data),

  deleteSegment: (segmentId: string) =>
    api.delete(`/transcription/segments/${segmentId}`).then((r) => r.data),

  mergeSegments: (segmentIds: string[]) =>
    api
      .post<Segment>('/transcription/segments/merge', { segment_ids: segmentIds })
      .then((r) => r.data),

  splitSegment: (segmentId: string, splitTime: number) =>
    api
      .post<Segment[]>(`/transcription/segments/${segmentId}/split`, { split_time: splitTime })
      .then((r) => r.data),

  getModels: () =>
    api
      .get<{ models: WhisperModel[]; recommended: string }>('/transcription/models')
      .then((r) => r.data),
}
